pkg/ipcache/ipcache.go

```diff
@@ -133,13 +160,20 @@ func (ipc *IPCache) SetListeners(listeners []IPIdentityMappingListener) {
 	ipc.mutex.Unlock()
 }
 
-// AddListenerLocked adds a listener for this IPCache.
+// AddListener adds a listener for this IPCache.
 func (ipc *IPCache) AddListener(listener IPIdentityMappingListener) {
-	ipc.mutex.Lock()
+	// We need to acquire the semaphore with a weight of lockWeight because we
+	// are modifying the ipc.listeners
+	ipc.mutex.Acquire(context.Background(), lockWeight)
 	ipc.listeners = append(ipc.listeners, listener)
+	// We will release the semaphore with a weight of rLock, *and not lockWeight*
+	// because want to prevent a race across an Upsert or Delete. By doing this
+	// we are sure no other writers are performing any operation while we are
+	// still reading.
+	ipc.mutex.Release(rLockWeight)
```

Comment:
You want to decrement by `(lockWeight-rlockWeight)` so you can allow a bigger number of readers.